src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -319,6 +323,65 @@ def _ensure_signature_valid(self, node: foast.Call, **kwargs) -> None:
                 node, msg=f"Invalid argument types in call to `{node.func.id}`!"
             ) from err
 
+    def _visit_math_built_in(self, node: foast.Call, **kwargs) -> foast.Call:
+        func_name = node.func.id
+
+        # validate arguments
+        error_msg_preamble = f"Incompatible argument in call to `{func_name}`."
+        if func_name in fbuiltins.UNARY_MATH_NUMBER_BUILTIN_NAMES:
+            arg_validator = type_info.is_arithmetic
+        elif func_name in fbuiltins.UNARY_MATH_FP_BUILTIN_NAMES:
+            arg_validator = type_info.is_floating_point
+        elif func_name in fbuiltins.UNARY_MATH_FP_PREDICATE_BUILTIN_NAMES:
+            arg_validator = type_info.is_floating_point
+        elif func_name in fbuiltins.BINARY_MATH_NUMBER_BUILTIN_NAMES:
+            arg_validator = type_info.is_arithmetic
```

Comment:
This introduces overloads for `power` and `mod` that also support integer types. This is technically fine. However, C++ also has these overloads, but will cast them to `double` in this case and return a double value (see below). Whereas, numpy seems to preserve integer types (see `numpy.power` vs `numpy.float_power`). I'm just unsure how easy/hard it would be to have a CUDA `power` function that preserves integer types. It might be easier to only support floating point values for now.

From: https://en.cppreference.com/w/cpp/numeric/math/pow
> If any argument has [integral type](https://en.cppreference.com/w/cpp/types/is_integral), it is cast to double

Reply:
A power implementation for integer types in cuda should be rather easy or am I missing something? After all we will need to wrap the the built-ins anyway to make sure we adhere to numpy semantics. I would like to keep the functionality as otherwise users start to write code like `some_int**2.`.